graph/src/components/ethereum/network.rs

```diff
@@ -0,0 +1,273 @@
+use std::cmp::Ord;
+use std::collections::{BTreeMap, HashMap};
+use std::fmt;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use crate::components::ethereum::EthereumAdapter;
+pub use crate::impl_slog_value;
+
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum NetworkCapability {
+    Full,
+    Archive,
+    Traces,
+}
+
+impl FromStr for NetworkCapability {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "full" => NetworkCapability::Full,
+            "archive" => NetworkCapability::Archive,
+            "traces" => NetworkCapability::Traces,
+            s => return Err(format_err!("invalid network capability provided: {}", s)),
+        })
+    }
+}
+
+impl fmt::Display for NetworkCapability {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            NetworkCapability::Full => write!(f, "full"),
+            NetworkCapability::Archive => write!(f, "archive"),
+            NetworkCapability::Traces => write!(f, "traces"),
+        }
+    }
+}
+
+impl NetworkCapability {
+    pub fn sufficient_capability(
+        network: &Vec<NetworkCapability>,
+        required: &Vec<NetworkCapability>,
+    ) -> bool {
+        match required[..] {
+            [NetworkCapability::Full] => true,
+            [NetworkCapability::Traces] => network.contains(&NetworkCapability::Traces),
+            [NetworkCapability::Full, NetworkCapability::Traces] => {
+                network.contains(&NetworkCapability::Traces)
+            }
+            [NetworkCapability::Archive] => network.contains(&NetworkCapability::Archive),
+            [NetworkCapability::Archive, NetworkCapability::Traces] => {
+                network == &vec![NetworkCapability::Archive, NetworkCapability::Traces]
+            }
+            _ => false,
+        }
+    }
+
+    pub fn cheapest() -> Self {
+        NetworkCapability::Full
+    }
+}
+
+impl_slog_value!(NetworkCapability, "{:?}");
+
+#[derive(Clone)]
+pub struct EthereumNetworkAdapters {
+    pub adapters: BTreeMap<Vec<NetworkCapability>, Arc<dyn EthereumAdapter>>,
+}
+
+impl EthereumNetworkAdapters {
+    pub fn cheapest_with(
+        &self,
+        required_capabilities: &Vec<NetworkCapability>,
+    ) -> Result<&Arc<dyn EthereumAdapter>, Error> {
+        let sufficient_adapters: BTreeMap<&Vec<NetworkCapability>, &Arc<dyn EthereumAdapter>> =
+            self.adapters
+                .iter()
+                .filter(|(capabilities, _adapter)| {
+                    NetworkCapability::sufficient_capability(capabilities, required_capabilities)
+                })
+                .collect();
+        if sufficient_adapters.len() == 0 {
+            return Err(format_err!(
+                "A matching Ethereum network with {:?} was not found.",
+                required_capabilities
+            ));
+        }
+        Ok(sufficient_adapters.values().next().unwrap())
```

Comment:
With the above construction (simple vector of adapters), this can be simplified to something like
```rust
let min_required = required_capabilities.iter().max();

let capable_adapters = self.adapters
    .filter(|adapter|  adapter.capabilities.iter().max() >= min_required)
    .collect();

if capable_adapters.is_empty() {
  return Err(...)
} else {
  // pick a random adapter from `capable_adapters` and return it
}
```

Reply:
Everything else in this file becomes simpler as well.